File: backend/quant/quant_engine.py
#!/usr/bin/env python3
"""Small, dependency-light quantitative research engine.

Input: JSON on stdin with a list of {timestamp, price} observations and strategy params.
Output: JSON containing equity curve, trades, and risk/performance statistics.

The engine deliberately shifts positions by one observation so signals are not
executed using the same close that created the signal (avoids simple look-ahead bias).
"""
import json
import math
import statistics
import sys
from typing import List, Dict, Tuple


def mean(xs):
    return sum(xs) / len(xs) if xs else 0.0


def stdev(xs):
    if len(xs) < 2:
        return 0.0
    return statistics.stdev(xs)


def returns(prices):
    out = [0.0]
    for i in range(1, len(prices)):
        prev = prices[i - 1]
        out.append((prices[i] / prev - 1.0) if prev else 0.0)
    return out


def rolling_stats(values, window):
    means, stds = [], []
    for i in range(len(values)):
        w = values[max(0, i - window + 1): i + 1]
        means.append(mean(w))
        stds.append(stdev(w))
    return means, stds


def momentum_signal(prices, lookback=10):
    sig = [0] * len(prices)
    for i in range(lookback, len(prices)):
        if prices[i] > prices[i - lookback]:
            sig[i] = 1
        elif prices[i] < prices[i - lookback]:
            sig[i] = -1
    return sig


def mean_reversion_signal(prices, window=20, entry_z=1.5, exit_z=0.25):
    means, stds = rolling_stats(prices, window)
    sig = [0] * len(prices)
    position = 0
    for i in range(len(prices)):
        if stds[i] <= 0:
            sig[i] = position
            continue
        z = (prices[i] - means[i]) / stds[i]
        if z <= -entry_z:
            position = 1
        elif z >= entry_z:
            position = -1
        elif abs(z) <= exit_z:
            position = 0
        sig[i] = position
    return sig


def max_drawdown(equity):
    if not equity:
        return 0.0
    peak = equity[0]
    worst = 0.0
    for x in equity:
        peak = max(peak, x)
        if peak:
            worst = min(worst, x / peak - 1.0)
    return worst


def performance_metrics(strategy_returns, equity, turnovers, annualization=252):
    r = strategy_returns[1:] if len(strategy_returns) > 1 else []
    if not r:
        return {}
    mu = mean(r)
    vol = stdev(r)
    sharpe = (mu / vol * math.sqrt(annualization)) if vol > 0 else 0.0
    downside = [x for x in r if x < 0]
    downside_vol = stdev(downside) if len(downside) > 1 else 0.0
    sortino = (mu / downside_vol * math.sqrt(annualization)) if downside_vol > 0 else 0.0
    sorted_r = sorted(r)
    idx = max(0, min(len(sorted_r) - 1, math.floor(0.05 * len(sorted_r))))
    var95 = sorted_r[idx]
    tail = [x for x in r if x <= var95]
    es95 = mean(tail) if tail else var95
    total_return = equity[-1] - 1.0 if equity else 0.0
    return {
        "total_return": total_return,
        "annualized_return": (1.0 + total_return) ** (annualization / max(1, len(r))) - 1.0,
        "volatility": vol * math.sqrt(annualization),
        "sharpe": sharpe,
        "sortino": sortino,
        "max_drawdown": max_drawdown(equity),
        "var_95": var95,
        "expected_shortfall_95": es95,
        "average_daily_return": mu,
        "turnover": sum(turnovers),
        "observations": len(r),
    }


def run_backtest(payload):
    observations = payload.get("data", [])
    strategy = payload.get("strategy", "mean_reversion")
    initial_capital = float(payload.get("initial_capital", 100000.0))
    fee_bps = float(payload.get("fee_bps", 5.0))
    slippage_bps = float(payload.get("slippage_bps", 2.0))
    leverage = float(payload.get("leverage", 1.0))

    observations = sorted(observations, key=lambda x: x.get("timestamp", ""))
    prices = [float(x["price"]) for x in observations if float(x.get("price", 0)) > 0]
    timestamps = [x.get("timestamp", i) for i, x in enumerate(observations) if float(x.get("price", 0)) > 0]
    if len(prices) < 5:
        raise ValueError("At least 5 valid price observations are required")

    params = payload.get("params", {})
    if strategy == "momentum":
        signals = momentum_signal(prices, int(params.get("lookback", 10)))
    elif strategy == "mean_reversion":
        signals = mean_reversion_signal(
            prices,
            int(params.get("window", 20)),
            float(params.get("entry_z", 1.5)),
            float(params.get("exit_z", 0.25)),
        )
    else:
        raise ValueError("Unknown strategy. Use momentum or mean_reversion.")

    # Position held during interval i -> i+1 uses signal[i], preventing same-bar lookahead.
    equity = [1.0]
    strategy_returns = [0.0]
    turnovers = [0.0]
    trades = []
    current_position = 0
    total_cost = 0.0

    for i in range(len(prices) - 1):
        target = max(-1, min(1, signals[i])) * leverage
        turnover = abs(target - current_position)
        gross = target * (prices[i + 1] / prices[i] - 1.0)
        trading_cost = turnover * (fee_bps + slippage_bps) / 10000.0
        net = gross - trading_cost
        equity.append(equity[-1] * (1.0 + net))
        strategy_returns.append(net)
        turnovers.append(turnover)
        total_cost += trading_cost

        if target != current_position:
            trades.append({
                "timestamp": timestamps[i],
                "price": prices[i],
                "from_position": current_position,
                "to_position": target,
                "turnover": turnover,
                "estimated_cost": trading_cost,
            })
            current_position = target

    metrics = performance_metrics(strategy_returns, equity, turnovers)
    metrics["estimated_transaction_cost"] = total_cost
    metrics["final_capital"] = initial_capital * equity[-1]
    metrics["pnl"] = initial_capital * (equity[-1] - 1.0)
    metrics["trade_count"] = len(trades)

    curve = [
        {
            "timestamp": timestamps[i],
            "price": prices[i],
            "equity": initial_capital * equity[i],
            "return": strategy_returns[i],
        }
        for i in range(len(prices))
    ]

    return {
        "strategy": strategy,
        "parameters": params,
        "metrics": metrics,
        "equity_curve": curve,
        "trades": trades,
    }


def main():
    try:
        payload = json.load(sys.stdin)
        print(json.dumps(run_backtest(payload)))
    except Exception as exc:
        print(json.dumps({"error": str(exc)}))
        sys.exit(1)


if __name__ == "__main__":
    main()

File: backend/quant/cpp/quant_engine.cpp
#include <algorithm>
#include <cmath>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <numeric>
#include <sstream>
#include <string>
#include <vector>

// Standalone C++ quantitative research engine.
// CSV format: timestamp,price
// Usage: ./quant_engine prices.csv momentum 10
//     or ./quant_engine prices.csv mean_reversion 20 1.5 0.25

struct Bar { std::string timestamp; double price; };

static double mean(const std::vector<double>& x, size_t l, size_t r) {
    if (r <= l) return 0.0;
    return std::accumulate(x.begin()+l, x.begin()+r, 0.0) / double(r-l);
}

static double stdev(const std::vector<double>& x, size_t l, size_t r) {
    if (r <= l + 1) return 0.0;
    double m = mean(x,l,r), s = 0.0;
    for (size_t i=l;i<r;i++) s += (x[i]-m)*(x[i]-m);
    return std::sqrt(s / double(r-l-1));
}

int main(int argc, char** argv) {
    if (argc < 4) {
        std::cerr << "Usage: quant_engine <csv> <momentum|mean_reversion> <window> [entry_z] [exit_z]\n";
        return 1;
    }

    std::ifstream in(argv[1]);
    if (!in) { std::cerr << "Cannot open CSV\n"; return 1; }

    std::vector<Bar> bars;
    std::string line;
    std::getline(in,line); // header
    while (std::getline(in,line)) {
        std::stringstream ss(line);
        std::string ts, ps;
        if (!std::getline(ss,ts,',')) continue;
        if (!std::getline(ss,ps,',')) continue;
        try { bars.push_back({ts,std::stod(ps)}); } catch (...) {}
    }
    if (bars.size() < 5) { std::cerr << "Need at least 5 observations\n"; return 1; }

    const std::string strategy = argv[2];
    const int window = std::max(1, std::stoi(argv[3]));
    const double entry = argc > 4 ? std::stod(argv[4]) : 1.5;
    const double exit = argc > 5 ? std::stod(argv[5]) : 0.25;

    std::vector<double> p;
    for (auto& b: bars) if (b.price > 0) p.push_back(b.price);
    std::vector<int> signal(p.size(),0);
    int position = 0;

    for (size_t i=0;i<p.size();i++) {
        if (strategy == "momentum") {
            if (i >= (size_t)window) signal[i] = p[i] > p[i-window] ? 1 : (p[i] < p[i-window] ? -1 : 0);
        } else if (strategy == "mean_reversion") {
            size_t l = i >= (size_t)window-1 ? i-(window-1) : 0;
            double m = mean(p,l,i+1), sd = stdev(p,l,i+1);
            if (sd > 0) {
                double z = (p[i]-m)/sd;
                if (z <= -entry) position = 1;
                else if (z >= entry) position = -1;
                else if (std::abs(z) <= exit) position = 0;
            }
            signal[i] = position;
        } else {
            std::cerr << "Unknown strategy\n"; return 1;
        }
    }

    double equity = 1.0;
    double peak = 1.0, maxdd = 0.0;
    double sum = 0.0, sumsq = 0.0;
    int n = 0, trades = 0, prev = 0;
    const double cost_bps = 7.0;

    std::cout << "timestamp,equity,return,position\n";
    std::cout << std::fixed << std::setprecision(8);
    std::cout << bars[0].timestamp << ",1.00000000,0.00000000,0\n";

    for (size_t i=0;i+1<p.size();i++) {
        int target = signal[i];
        double turnover = std::abs(target-prev);
        double ret = target * (p[i+1]/p[i]-1.0) - turnover*cost_bps/10000.0;
        equity *= (1.0+ret);
        sum += ret; sumsq += ret*ret; n++;
        if (target != prev) trades++;
        prev = target;
        peak = std::max(peak,equity);
        maxdd = std::min(maxdd,equity/peak-1.0);
        std::cout << bars[i+1].timestamp << ',' << equity << ',' << ret << ',' << target << '\n';
    }

    double mu = n ? sum/n : 0.0;
    double var = n > 1 ? (sumsq - n*mu*mu)/(n-1) : 0.0;
    double vol = std::sqrt(std::max(0.0,var));
    double sharpe = vol > 0 ? mu/vol*std::sqrt(252.0) : 0.0;

    std::cerr << "Final equity: " << equity << '\n';
    std::cerr << "Total return: " << equity-1.0 << '\n';
    std::cerr << "Volatility: " << vol*std::sqrt(252.0) << '\n';
    std::cerr << "Sharpe: " << sharpe << '\n';
    std::cerr << "Max drawdown: " << maxdd << '\n';
    std::cerr << "Trades: " << trades << '\n';
    return 0;
}
